rclcpp_action/test/test_client.cpp

```diff
@@ -14,29 +14,413 @@
+  EXPECT_FALSE(goal_handle->is_feedback_aware());
+  EXPECT_TRUE(goal_handle->is_result_aware());
+  auto future_result = goal_handle->async_result();
+  dual_spin_until_future_complete(future_result);
+  auto result = future_result.get();
+  ASSERT_EQ(6ul, result.response->sequence.size());
+  EXPECT_EQ(0, result.response->sequence[0]);
+  EXPECT_EQ(1, result.response->sequence[1]);
+  EXPECT_EQ(5, result.response->sequence[5]);
+}
+
+TEST_F(TestClient, async_send_goal_with_feedback_and_result)
+{
+  auto action_client = rclcpp_action::create_client<ActionType>(client_node, action_name);
+
+  ActionGoal goal;
+  goal.order = 4;
+  int feedback_count = 0;
+  auto future_goal_handle = action_client->async_send_goal(
+    goal,
+    [&feedback_count](
+      typename ActionGoalHandle::SharedPtr goal_handle,
+      const std::shared_ptr<const ActionFeedback> feedback) mutable
+    {
+      (void)goal_handle;
+      (void)feedback;
+      feedback_count++;
+    });
+  dual_spin_until_future_complete(future_goal_handle);
+  auto goal_handle = future_goal_handle.get();
+  EXPECT_EQ(rclcpp_action::GoalStatus::STATUS_ACCEPTED, goal_handle->get_status());
+  EXPECT_TRUE(goal_handle->is_feedback_aware());
+  EXPECT_TRUE(goal_handle->is_result_aware());
+  auto future_result = goal_handle->async_result();
+  dual_spin_until_future_complete(future_result);
+  auto result = future_result.get();
+
+  ASSERT_EQ(5u, result.response->sequence.size());
+  EXPECT_EQ(3, result.response->sequence.back());
+  EXPECT_EQ(5, feedback_count);
+}
+
+TEST_F(TestClient, async_cancel_one_goal)
+{
+  auto action_client = rclcpp_action::create_client<ActionType>(client_node, action_name);
+
+  ActionGoal goal;
+  goal.order = 5;
+  auto future_goal_handle = action_client->async_send_goal(goal, nullptr, true);
+  dual_spin_until_future_complete(future_goal_handle);
+  auto goal_handle = future_goal_handle.get();
+  EXPECT_EQ(rclcpp_action::GoalStatus::STATUS_ACCEPTED, goal_handle->get_status());
+
+  auto future_cancel = action_client->async_cancel_goal(goal_handle);
+  dual_spin_until_future_complete(future_cancel);
+  bool goal_canceled = future_cancel.get();
+  EXPECT_TRUE(goal_canceled);
+}
+
+TEST_F(TestClient, async_cancel_all_goals)
+{
+  auto action_client = rclcpp_action::create_client<ActionType>(client_node, action_name);
+
+  ActionGoal goal;
+  goal.order = 6;
+  auto future_goal_handle0 = action_client->async_send_goal(goal, nullptr, true);
+  dual_spin_until_future_complete(future_goal_handle0);
+  auto goal_handle0 = future_goal_handle0.get();
+
+  ASSERT_EQ(RCL_RET_OK, rcl_set_ros_time_override(clock.get_clock_handle(), RCL_S_TO_NS(2)));
+
+  goal.order = 8;
+  auto future_goal_handle1 = action_client->async_send_goal(goal, nullptr, true);
+  dual_spin_until_future_complete(future_goal_handle1);
+  auto goal_handle1 = future_goal_handle1.get();
+
+  if (goal_handle1->get_goal_id() < goal_handle0->get_goal_id()) {
+    goal_handle0.swap(goal_handle1);
```

Comment:
To clarify, is this just so we can know the order when checking UUIDs below (L395) ?